In [1]:
import pandas as pd 
import os
from Functions import evaluate

In [2]:
os.listdir("./data/evaluate/raw/")

['.ipynb_checkpoints',
 'cast.csv',
 'clean_movies.csv',
 'crew.csv',
 'genres.csv',
 'keywords.csv',
 'prod_company.csv',
 'userrating.csv']

In [3]:
movies = pd.read_csv('./data/evaluate/raw/clean_movies.csv')

In [4]:
movies.head(2)

id  adult original_title  runtime  \
0   862  False      Toy Story     81.0   
1  8844  False        Jumanji    104.0   

                                              genres  popularity  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   21.946943   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   17.015539   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                    release_date      revenue  vote_average  
0  1995-10-30 00:00:00.000000000  373554033.0           7.7  
1  1995-12-15 00:00:00.000000000  262797249.0           6.9

In [5]:
movies.columns

Index(['id', 'adult', 'original_title', 'runtime', 'genres', 'popularity',
       'production_companies', 'release_date', 'revenue', 'vote_average'],
      dtype='object')

In [6]:
cols_needed = ['id', 'adult', 'original_title', 'runtime',
               'popularity','revenue', 'vote_average','release_date']
movies = movies[cols_needed]

In [7]:
movies.head(2)

id  adult original_title  runtime  popularity      revenue  vote_average  \
0   862  False      Toy Story     81.0   21.946943  373554033.0           7.7   
1  8844  False        Jumanji    104.0   17.015539  262797249.0           6.9   

                    release_date  
0  1995-10-30 00:00:00.000000000  
1  1995-12-15 00:00:00.000000000

In [8]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45446 entries, 0 to 45445
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              45446 non-null  int64  
 1   adult           45446 non-null  bool   
 2   original_title  45446 non-null  object 
 3   runtime         45446 non-null  float64
 4   popularity      45446 non-null  float64
 5   revenue         45446 non-null  float64
 6   vote_average    45446 non-null  float64
 7   release_date    45446 non-null  object 
dtypes: bool(1), float64(4), int64(1), object(2)
memory usage: 2.5+ MB


In [9]:
movies.rename(columns={"id" : 'movie_id'}, inplace=True)

In [10]:
movies['release_date'] = pd.to_datetime(movies.release_date)

## Add Dim ID

In [11]:
os.listdir("./data/out/")

['.ipynb_checkpoints',
 'box_office_data.csv',
 'Dim-Cast.csv',
 'Dim-Crew.csv',
 'Dim-Dates.csv',
 'Dim-genres.csv',
 'Dim-keywords.csv',
 'Dim-Production-company.csv',
 'Fact.csv']

In [12]:
cast = pd.read_csv('./data/out/Dim-Cast.csv')
cast.columns

Index(['character', 'name', 'order', 'movie_id', 'DimCastID'], dtype='object')

In [13]:
movies = movies.merge(cast[['movie_id','DimCastID']] , how='left', on = 'movie_id').drop_duplicates()

In [14]:
crew = pd.read_csv('./data/out/Dim-Crew.csv')
crew.columns

Index(['name', 'department', 'job', 'movie_id', 'DimCrewID'], dtype='object')

In [15]:
movies = movies.merge(crew[['movie_id','DimCrewID']] , how='left', on = 'movie_id').drop_duplicates()

In [16]:
dates = pd.read_csv('./data/out/Dim-Dates.csv')
dates.rename(columns={"id" : "movie_id"},inplace=True)
dates.columns

Index(['movie_id', 'release_date', 'year', 'quarter', 'month', 'day',
       'DimDateID'],
      dtype='object')

In [17]:
movies = movies.merge(dates[['movie_id','DimDateID']] , how='left', on = 'movie_id').drop_duplicates()

In [18]:
genre = pd.read_csv('./data/out/Dim-genres.csv')
genre.columns

Index(['name', 'movie_id', 'DimGenreID'], dtype='object')

In [19]:
movies = movies.merge(genre[['movie_id','DimGenreID']] , how='left', on = 'movie_id').drop_duplicates()

In [20]:
keywords = pd.read_csv('./data/out/Dim-keywords.csv')
keywords.columns

Index(['name', 'movie_id', 'DimKeywordsID'], dtype='object')

In [21]:
movies = movies.merge(keywords[['movie_id','DimKeywordsID']] , how='left', on = 'movie_id').drop_duplicates()

In [22]:
pc = pd.read_csv('./data/out/Dim-Production-company.csv')
pc.columns

Index(['name', 'movie_id', 'DimProdCompID'], dtype='object')

In [23]:
movies = movies.merge(pc[['movie_id','DimProdCompID']] , how='left', on = 'movie_id').drop_duplicates()

In [24]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45446 entries, 0 to 82372
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   movie_id        45446 non-null  int64         
 1   adult           45446 non-null  bool          
 2   original_title  45446 non-null  object        
 3   runtime         45446 non-null  float64       
 4   popularity      45446 non-null  float64       
 5   revenue         45446 non-null  float64       
 6   vote_average    45446 non-null  float64       
 7   release_date    45446 non-null  datetime64[ns]
 8   DimCastID       43031 non-null  float64       
 9   DimCrewID       44674 non-null  float64       
 10  DimDateID       45446 non-null  int64         
 11  DimGenreID      43004 non-null  float64       
 12  DimKeywordsID   31103 non-null  float64       
 13  DimProdCompID   33572 non-null  float64       
dtypes: bool(1), datetime64[ns](1), float64(9), int64(2), o

In [25]:
# Replace nan with -1 
# the Dim tables Contain ID -1 With Not Provided Data 

In [26]:
movies.fillna(-1, inplace=True)

### Add Box Office

In [27]:
box_office = pd.read_csv('data/out/box_office_data.csv')

In [28]:
box_office.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15918 entries, 0 to 15917
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   original_title   15918 non-null  object
 1   year             15918 non-null  int64 
 2   domestic_gross   15918 non-null  int64 
 3   overseas_gross   15918 non-null  int64 
 4   worldwide_gross  15918 non-null  int64 
dtypes: int64(4), object(1)
memory usage: 621.9+ KB


I merge to movies data and box office on left outer join and on keys {original_title,year}
first i need to extract year from release_date in movies data 

In [29]:
movies['year'] = pd.to_datetime(movies.release_date).dt.year

In [30]:
movies = movies.merge(box_office,how='left',on=['original_title','year'])

In [31]:
movies.fillna(0,inplace=True)

In [32]:
movies.query('revenue == 0 & worldwide_gross!=0').shape[0]

5044

In [33]:
movies.head(10)

movie_id  adult               original_title  runtime  popularity  \
0       862  False                    Toy Story     81.0   21.946943   
1      8844  False                      Jumanji    104.0   17.015539   
2     15602  False             Grumpier Old Men    101.0   11.712900   
3     31357  False            Waiting to Exhale    127.0    3.859495   
4     11862  False  Father of the Bride Part II    106.0    8.387519   
5       949  False                         Heat    170.0   17.924927   
6     11860  False                      Sabrina    127.0    6.677277   
7     45325  False                 Tom and Huck     97.0    2.561161   
8      9091  False                 Sudden Death    106.0    5.231580   
9       710  False                    GoldenEye    130.0   14.686036   

       revenue  vote_average release_date  DimCastID  DimCrewID  DimDateID  \
0  373554033.0           7.7   1995-10-30    16223.0      263.0          1   
1  262797249.0           6.9   1995-12-15     3674.0     7013.0          2   
2          0.0           6.5   1995-12-22    29060.0    25692.0          3   
3   81452156.0           6.1   1995-12-22    22681.0    14045.0          4   
4   76578911.0           5.7   1995-02-10    18414.0    20961.0          5   
5  187436818.0           7.7   1995-12-15      375.0      659.0          6   
6          0.0           6.2   1995-12-15      530.0     1059.0          7   
7          0.0           5.4   1995-12-22    29106.0    25750.0          8   
8   64350171.0           5.5   1995-12-22    32083.0    16952.0          9   
9  352194034.0           6.6   1995-11-16     6601.0     1328.0         10   

   DimGenreID  DimKeywordsID  DimProdCompID  year  domestic_gross  \
0      6801.0         4872.0        22856.0  1995     192364233.0   
1      5172.0         9189.0         5435.0  1995     100475249.0   
2     19431.0        15085.0        15661.0  1995      69870000.0   
3     13942.0        10481.0        26432.0  1995      67052156.0   
4     29138.0         5180.0        15659.0  1995      76594000.0   
5      3634.0          153.0         7149.0  1995      67436818.0   
6     19435.0         7886.0          845.0  1995      53672080.0   
7      3635.0           -1.0        26431.0  1995      23735802.0   
8     13859.0        15086.0         7147.0  1995      20350171.0   
9     13849.0         1580.0        15654.0  1995     106600000.0   

   overseas_gross  worldwide_gross  
0     181757800.0      374122033.0  
1     162322000.0      262797249.0  
2             0.0       69870000.0  
3      14400000.0       81452156.0  
4             0.0       76594000.0  
5     120000000.0      187436818.0  
6      33641681.0       87313761.0  
7             0.0       23735802.0  
8      44000000.0       64350171.0  
9     245764093.0      352364093.0

Nice we will fill 5k of movie with 0 revenue

In [34]:
# get number of movies that has 0 revenue
(movies.revenue == 0).sum()

38042

Replace 0 revenue with worldwide_gross

In [35]:
indx = movies.query('revenue == 0 & worldwide_gross!=0')['revenue'].index.tolist()
movies.loc[indx,['revenue']] = movies.loc[indx,['worldwide_gross']].values

In [36]:
(movies.revenue == 0).sum()

32998

In [37]:
movies.head()

movie_id  adult               original_title  runtime  popularity  \
0       862  False                    Toy Story     81.0   21.946943   
1      8844  False                      Jumanji    104.0   17.015539   
2     15602  False             Grumpier Old Men    101.0   11.712900   
3     31357  False            Waiting to Exhale    127.0    3.859495   
4     11862  False  Father of the Bride Part II    106.0    8.387519   

       revenue  vote_average release_date  DimCastID  DimCrewID  DimDateID  \
0  373554033.0           7.7   1995-10-30    16223.0      263.0          1   
1  262797249.0           6.9   1995-12-15     3674.0     7013.0          2   
2   69870000.0           6.5   1995-12-22    29060.0    25692.0          3   
3   81452156.0           6.1   1995-12-22    22681.0    14045.0          4   
4   76578911.0           5.7   1995-02-10    18414.0    20961.0          5   

   DimGenreID  DimKeywordsID  DimProdCompID  year  domestic_gross  \
0      6801.0         4872.0        22856.0  1995     192364233.0   
1      5172.0         9189.0         5435.0  1995     100475249.0   
2     19431.0        15085.0        15661.0  1995      69870000.0   
3     13942.0        10481.0        26432.0  1995      67052156.0   
4     29138.0         5180.0        15659.0  1995      76594000.0   

   overseas_gross  worldwide_gross  
0     181757800.0      374122033.0  
1     162322000.0      262797249.0  
2             0.0       69870000.0  
3      14400000.0       81452156.0  
4             0.0       76594000.0

In [38]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45467 entries, 0 to 45466
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   movie_id         45467 non-null  int64         
 1   adult            45467 non-null  bool          
 2   original_title   45467 non-null  object        
 3   runtime          45467 non-null  float64       
 4   popularity       45467 non-null  float64       
 5   revenue          45467 non-null  float64       
 6   vote_average     45467 non-null  float64       
 7   release_date     45467 non-null  datetime64[ns]
 8   DimCastID        45467 non-null  float64       
 9   DimCrewID        45467 non-null  float64       
 10  DimDateID        45467 non-null  int64         
 11  DimGenreID       45467 non-null  float64       
 12  DimKeywordsID    45467 non-null  float64       
 13  DimProdCompID    45467 non-null  float64       
 14  year             45467 non-null  int64

### Add Normalized Revenue Against a Customer Price Index (CPI)

In [39]:
cpi = pd.read_csv('data/in/fred/Consumer Price Index of Admission to Movies.csv')

In [40]:
cpi['date'] = pd.to_datetime(cpi.date)
cpi['year'] = pd.to_datetime(cpi.date).dt.year
cpi['month'] = pd.to_datetime(cpi.date).dt.month

In [41]:
cpi.head()

date  value  year  month
0 1999-01-01  102.7  1999      1
1 1999-02-01  103.7  1999      2
2 1999-03-01  104.2  1999      3
3 1999-04-01  104.8  1999      4
4 1999-05-01  106.4  1999      5

In [42]:
cpi.date.min(), cpi.date.max()

(Timestamp('1999-01-01 00:00:00'), Timestamp('2022-08-01 00:00:00'))

In [43]:
movies.query('release_date >= @cpi.date.min() & release_date <= @cpi.date.max()').head()

movie_id  adult     original_title  runtime  popularity  revenue  \
707     131232  False          Due Amici     86.0    0.003949      0.0   
868      79782  False            Wenecja    110.0    0.147130      0.0   
1081    141210  False      The Sleepover      6.0    0.135596      0.0   
1654    216794  False       Men of Means     80.0    0.022078      0.0   
2113    143750  False  The Farmer's Wife     18.0    0.211754      0.0   

      vote_average release_date  DimCastID  DimCrewID  DimDateID  DimGenreID  \
707            0.0   2002-03-20    40906.0    26029.0        708     19785.0   
868            7.5   2010-05-25    32050.0    24350.0        869     19846.0   
1081           8.0   2013-10-12    36222.0    41724.0       1082     18302.0   
1654           0.0   1999-05-25    36179.0    33685.0       1655     18422.0   
2113          10.0   2012-06-20    39124.0    17040.0       2114     30377.0   

      DimKeywordsID  DimProdCompID  year  domestic_gross  overseas_gross  \
707            -1.0           -1.0  2002             0.0             0.0   
868            -1.0           -1.0  2010             0.0             0.0   
1081           -1.0           -1.0  2013             0.0             0.0   
1654        19810.0        16206.0  1999             0.0             0.0   
2113        25106.0           -1.0  2012             0.0             0.0   

      worldwide_gross  
707               0.0  
868               0.0  
1081              0.0  
1654              0.0  
2113              0.0

In [44]:
(movies.query('release_date >= @cpi.date.min() & release_date <= @cpi.date.max()')['revenue']==0).sum()

17513

We can normalize revenue of 24742 movies and almost 17513 of them have 0 revenue 

In [45]:
def get_normalize(data):
    base_cpi = float(cpi['value'][-1:].values)
    m_revenue = data['revenue']
    m_year = int(pd.to_datetime(data.release_date).dt.year.values)
    m_month = int(pd.to_datetime(data.release_date).dt.month.values)
    movie_cpi = float(cpi.query('year == 1999 & month == 1')['value'].values)
    if m_year<1999 or m_year>2022:
        return 0
    else:
        return int(m_revenue * (base_cpi/movie_cpi))

In [46]:
normalize_rev = []
y = 1
for i in range(movies.shape[0]):
    normalize_rev.append(get_normalize(movies[i:y]))
    y+=1
    
movies['revenue_normalized_08_2022'] = normalize_rev

### convert Dim ID to Int

In [47]:
movies[['DimCastID','DimCrewID','DimGenreID','DimKeywordsID','DimProdCompID']] = \
movies[['DimCastID','DimCrewID','DimGenreID','DimKeywordsID','DimProdCompID']].astype('int64')

In [49]:
movies.columns

Index(['movie_id', 'adult', 'original_title', 'runtime', 'popularity',
       'revenue', 'vote_average', 'release_date', 'DimCastID', 'DimCrewID',
       'DimDateID', 'DimGenreID', 'DimKeywordsID', 'DimProdCompID', 'year',
       'domestic_gross', 'overseas_gross', 'worldwide_gross',
       'revenue_normalized_08_2022'],
      dtype='object')

In [51]:
movies = movies[['DimDateID','DimCastID','DimCrewID','DimProdCompID','DimKeywordsID','DimGenreID','movie_id',
        'original_title','runtime','release_date','popularity','vote_average','revenue','revenue_normalized_08_2022']]

In [52]:
movies.head()

DimDateID  DimCastID  DimCrewID  DimProdCompID  DimKeywordsID  DimGenreID  \
0          1      16223        263          22856           4872        6801   
1          2       3674       7013           5435           9189        5172   
2          3      29060      25692          15661          15085       19431   
3          4      22681      14045          26432          10481       13942   
4          5      18414      20961          15659           5180       29138   

   movie_id               original_title  runtime release_date  popularity  \
0       862                    Toy Story     81.0   1995-10-30   21.946943   
1      8844                      Jumanji    104.0   1995-12-15   17.015539   
2     15602             Grumpier Old Men    101.0   1995-12-22   11.712900   
3     31357            Waiting to Exhale    127.0   1995-12-22    3.859495   
4     11862  Father of the Bride Part II    106.0   1995-02-10    8.387519   

   vote_average      revenue  revenue_normalized_08_2022  
0           7.7  373554033.0                           0  
1           6.9  262797249.0                           0  
2           6.5   69870000.0                           0  
3           6.1   81452156.0                           0  
4           5.7   76578911.0                           0

In [54]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45467 entries, 0 to 45466
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   DimDateID                   45467 non-null  int64         
 1   DimCastID                   45467 non-null  int64         
 2   DimCrewID                   45467 non-null  int64         
 3   DimProdCompID               45467 non-null  int64         
 4   DimKeywordsID               45467 non-null  int64         
 5   DimGenreID                  45467 non-null  int64         
 6   movie_id                    45467 non-null  int64         
 7   original_title              45467 non-null  object        
 8   runtime                     45467 non-null  float64       
 9   release_date                45467 non-null  datetime64[ns]
 10  popularity                  45467 non-null  float64       
 11  vote_average                45467 non-null  float64   

#### save the data

In [53]:
movies.to_csv('data/out/Fact.csv',index=False)